eventmetrics/parsers/parserValidator.go

```diff
@@ -0,0 +1,191 @@
+			// default empty function
+		}
+	}
+}
+
+// log an cycle error to metrics
+func logCycleErr(err error, counter *prometheus.CounterVec, logger *zap.Logger) {
+	var taggedErrs validation.TaggedErrors
+	var taggedErr validation.TaggedError
+	if errors.As(err, &taggedErrs) {
+		logger.Info("invalid cycle", zap.String("tags", tagsToString(taggedErrs.UniqueTags())))
+		for _, tag := range taggedErrs.UniqueTags() {
+			counter.With(prometheus.Labels{reasonLabel: tag.String()}).Add(1.0)
+		}
+	} else if errors.As(err, &taggedErr) {
+		logger.Info("invalid cycle", zap.String("tags", taggedErr.Tag().String()))
+		counter.With(prometheus.Labels{reasonLabel: taggedErr.Tag().String()}).Add(1.0)
+	} else if err != nil {
+		logger.Info("invalid cycle; no tags", zap.Error(err))
+		counter.With(prometheus.Labels{reasonLabel: validation.Unknown.String()}).Add(1.0)
+	}
+}
+
+// log an event error to metrics
+func logEventError(logger *zap.Logger, counter *prometheus.CounterVec, err error, event interpreter.Event) {
+	const (
+		eventIDKey  = "event id"
+		deviceIDKey = "device id"
+	)
+
+	hardwareVal, firmwareVal, _ := getHardwareFirmware(event)
+	deviceID, _ := event.DeviceID()
+	eventID := event.TransactionUUID
+
+	var taggedErrs validation.TaggedErrors
+	var taggedErr validation.TaggedError
+	if errors.As(err, &taggedErrs) {
+		logger.Info("event validation error", zap.String("tags", tagsToString(taggedErrs.UniqueTags())), zap.String(eventIDKey, eventID), zap.String(deviceIDKey, deviceID))
+		for _, tag := range taggedErrs.UniqueTags() {
+			counter.With(prometheus.Labels{firmwareLabel: firmwareVal,
+				hardwareLabel: hardwareVal, reasonLabel: tag.String()}).Add(1.0)
+		}
+	} else if errors.As(err, &taggedErr) {
+		logger.Info("event validation error", zap.String("tags", taggedErr.Tag().String()), zap.String(eventIDKey, eventID), zap.String(deviceIDKey, deviceID))
+		counter.With(prometheus.Labels{firmwareLabel: firmwareVal,
+			hardwareLabel: hardwareVal, reasonLabel: taggedErr.Tag().String()}).Add(1.0)
+	} else if err != nil {
+		logger.Info("event validation error; no tags", zap.Error(err), zap.String(eventIDKey, eventID), zap.String(deviceIDKey, deviceID))
+		counter.With(prometheus.Labels{firmwareLabel: firmwareVal,
+			hardwareLabel: hardwareVal, reasonLabel: validation.Unknown.String()}).Add(1.0)
+	}
+}
+
+// get hardware and firmware values from event metadata, returning false if either one or both are not found
+func getHardwareFirmware(event interpreter.Event) (hardwareVal string, firmwareVal string, found bool) {
+	hardwareVal, hardwareFound := event.GetMetadataValue(hardwareMetadataKey)
+	firmwareVal, firmwareFound := event.GetMetadataValue(firmwareMetadataKey)
+
+	found = true
+	if !hardwareFound || !firmwareFound {
+		found = false
+		if !hardwareFound {
+			hardwareVal = unknownReason
+		}
+
+		if !firmwareFound {
+			firmwareVal = unknownReason
+		}
+	}
+
+	return
+}
+
+func tagsToString(tags []validation.Tag) string {
```

Comment:
I think this function is a bit much....As far as I can tell, this is only used for logging, so my preference would be to have this function convert `[]Tag` to `[]string` and then use [`zap.Strings()`](https://pkg.go.dev/go.uber.org/zap#Strings) to log the resulting slice.

This is probably a fairly common ask for `Tag`s so it might be better to export this helper function as a part of the `validation` package itself.